scipy/sparse/_compressed.py

```diff
@@ -366,105 +362,127 @@ def multiply(self, other):
-            # Row vector times matrix. other is a row.
-            elif other.shape[0] == 1 and self.shape[1] == other.shape[1]:
-                other = self._dia_container(
-                    (other.toarray().ravel(), [0]),
-                    shape=(other.shape[1], other.shape[1])
-                )
-                return self._matmul_sparse(other)
+            if sM == 1 and oN == 1:
+                return other._matmul_sparse(self.reshape(sM, sN).tocsc())
+            if sN == 1 and oM == 1:
+                return self._matmul_sparse(other.reshape(oM, oN).tocsc())
+
+            # Other is a row.
+            if oM == 1 and sN == oN:
+                # build diagonal csc_array/csr_array => self._csr_container
+                data = other.toarray().ravel()
+                indptr = np.arange(oN + 1)
+                indices = indptr[:-1]
+                new_other = self._csr_container((data, indices, indptr))
+                result = self._matmul_sparse(new_other)
+                return result if self.ndim == 2 else result.reshape((1, oN))
             # self is a row.
-            elif self.shape[0] == 1 and self.shape[1] == other.shape[1]:
-                copy = self._dia_container(
-                    (self.toarray().ravel(), [0]),
-                    shape=(self.shape[1], self.shape[1])
-                )
+            if sM == 1 and sN == oN:
+                data = self.toarray().ravel()
+                indptr = np.arange(oN + 1)
+                indices = indptr[:-1]
+                copy = self._csr_container((data, indices, indptr))
                 return other._matmul_sparse(copy)
-            # Column vector times matrix. other is a column.
-            elif other.shape[1] == 1 and self.shape[0] == other.shape[0]:
-                other = self._dia_container(
-                    (other.toarray().ravel(), [0]),
-                    shape=(other.shape[0], other.shape[0])
-                )
-                return other._matmul_sparse(self)
+
+            # Other is a column.
+            if oN == 1 and sM == oM:
+                data = other.toarray().ravel()
+                indptr = np.arange(oM + 1)
+                indices = indptr[:-1]
+                new_other = self._csr_container((data, indices, indptr))
+                return new_other._matmul_sparse(self)
             # self is a column.
-            elif self.shape[1] == 1 and self.shape[0] == other.shape[0]:
-                copy = self._dia_container(
-                    (self.toarray().ravel(), [0]),
-                    shape=(self.shape[0], self.shape[0])
-                )
-                return copy._matmul_sparse(other)
-            else:
-                raise ValueError("inconsistent shapes")
+            if sN == 1 and sM == oM:
+                data = self.toarray().ravel()
+                indptr = np.arange(oM + 1)
+                indices = indptr[:-1]
+                new_self = self._csr_container((data, indices, indptr))
```

Comment:
It might be useful to extract this logic into a helper function, since we're calling it 4 times here. Maybe something like `_diagonal_csr(data, size)`.